calculation/sha_calc/sha_calc/directivity/bea20/distributions.py

```diff
@@ -0,0 +1,314 @@
+                    planes_depth_copy = [plane.copy() for plane in planes]
+                    planes_depth_copy[index]["shyp"] = distance - (total_length / 2)
+                    planes_depth_copy[index]["dhyp"] = plane["width"] * depth
+                    planes_index.append(index)
+                    planes_list.append(planes_depth_copy)
+            current_length += plane["length"]
+
+    return planes_list, planes_index, []
+
+
+def even_grid(
+    hypo_along_strike: int,
+    hypo_down_dip: int,
+    planes: List,
+    event_type: EventType,
+    total_length: float,
+):
+    """
+    Uses an evenly spaced grid across strike and dip based on the amount of hypocentres specified across both lengths
+
+    Parameters
+    ----------
+    hypo_along_strike: int
+        Number of hypocentres across strike
+    hypo_down_dip: int
+        Number of hypocentres down dip
+    planes: list
+        The planes to adjust and set the hypocentre on
+    event_type: EventType
+        The event type Strike_slip, dip_slip or all for determining the down dip distribution function
+    total_length: float
+        The total length of the fault
+    """
+
+    gap = total_length / (hypo_along_strike + 1)
+    distances = [gap * (x + 1) for x in range(hypo_along_strike)]
+
+    down_gap = planes[0]["width"] / (hypo_down_dip + 1)
+    down_dip_values = [down_gap * (x + 1) for x in range(hypo_down_dip)]
+
+    # Works out the distances across strike of the fault for each hypocentre
+    # Based on a normal distribution and truncated between 0 and 1
+    mu, sigma = 0.5, 0.23
+    lower, upper = 0, 1
+    strike_distribution = stats.truncnorm(
+        (lower - mu) / sigma, (upper - mu) / sigma, loc=mu, scale=sigma
+    )
+    strike_weight_dist = strike_distribution.pdf(
+        [x / total_length for x in distances]
+    ) * (1 / hypo_along_strike)
+
+    # Works out the depth method for down dip placement of hypocentres
+    # Based on Weilbull or Gamma distributions depending on the EventType
+    if event_type == EventType.DIP_SLIP:
+        down_dip_distribution = gamma_truncated(a=0, b=1)
+    elif event_type == EventType.STRIKE_SLIP:
+        down_dip_distribution = weibull_truncated_strike_slip(a=0, b=1)
+    else:
+        down_dip_distribution = weibull_truncated_oblique(a=0, b=1)
+    dip_weight_dist = down_dip_distribution.pdf(
+        [x / planes[0]["width"] for x in down_dip_values]
+    ) * (1 / hypo_down_dip)
+
+    combo_weight_dist = [
+        strike * dip for strike in strike_weight_dist for dip in dip_weight_dist
+    ]
+    sum_dist = sum(combo_weight_dist)
+    weights = [x / sum_dist for x in combo_weight_dist]
+
+    planes_list = []
+    planes_index = []
+
+    for distance in distances:
+        current_length = 0
+        planes_copy = [plane.copy() for plane in planes]
+        for index, plane in enumerate(planes_copy):
+            if current_length < distance < (current_length + plane["length"]):
```

Comment:
Assuming my understanding is correct, this for loop is for selecting the correct plane given a strike location. If thats the case, then I think this can be simplified, with the following two lines:

`plane_lenghts = np.asarray([cur_plane["length"] for cur_plane in planes])`
before the distance for loop, just an array with the lenghts of all the planes

`plane_ix = np.flatnonzero(distance < np.cumsum(plane_lenghts))[0]`
which returns the index of the plane to use given the current strike location (i.e. distance)

I personally think this is cleaner and clearer (especially with a comment above the 2nd statement), but up to you